packages/firestore/test/integration/api/snasphot_listener_source.test.ts

```diff
@@ -0,0 +1,778 @@
+            { k: 'd', sort: -1 },
+            { k: 'a', sort: 0 }
+          ]);
+          expect(snapshot.metadata.hasPendingWrites).to.equal(true);
+
+          snapshot = await storeLimitToLastEvent.awaitEvent();
+          expect(toDataArray(snapshot)).to.deep.equal([
+            { k: 'a', sort: 0 },
+            { k: 'd', sort: -1 }
+          ]);
+          expect(snapshot.metadata.hasPendingWrites).to.equal(true);
+
+          // Un-listen to limitToLast, update a doc, then re-listen limitToLast.
+          limitToLastUnlisten();
+
+          await updateDoc(doc(coll, 'a'), { k: 'a', sort: -2 });
+          limitToLastUnlisten = onSnapshot(
+            query(coll, orderBy('sort', 'desc'), limitToLast(2)),
+            { source: ListenSource.Cache },
+            storeLimitToLastEvent.storeEvent
+          );
+
+          // Verify both queries get expected results.
+          snapshot = await storeLimitEvent.awaitEvent();
+          expect(toDataArray(snapshot)).to.deep.equal([
+            { k: 'a', sort: -2 },
+            { k: 'd', sort: -1 }
+          ]);
+          expect(snapshot.metadata.hasPendingWrites).to.equal(true);
+
+          snapshot = await storeLimitToLastEvent.awaitEvent();
+          expect(toDataArray(snapshot)).to.deep.equal([
+            { k: 'd', sort: -1 },
+            { k: 'a', sort: -2 }
+          ]);
+          // We listened to LimitToLast query after the doc update.
+          expect(snapshot.metadata.hasPendingWrites).to.equal(false);
+
+          limitUnlisten();
+          limitToLastUnlisten();
+        });
+      });
+
+      it('can listen to default source first and then cache', () => {
+        const testDocs = {
+          a: { k: 'a', sort: 0 }
+        };
+        return withTestCollection(persistence, testDocs, async coll => {
+          // Listen to the query with default options, which will also populates the cache
+          const storeDefaultEvent = new EventsAccumulator<QuerySnapshot>();
+          const defaultUnlisten = onSnapshot(
+            query(coll, orderBy('sort', 'asc')),
+            storeDefaultEvent.storeEvent
+          );
+          let snapshot = await storeDefaultEvent.awaitRemoteEvent();
+          expect(toDataArray(snapshot)).to.deep.equal([{ k: 'a', sort: 0 }]);
+          expect(snapshot.metadata.fromCache).to.equal(false);
+
+          // Listen to the same query from cache
+          const storeCacheEvent = new EventsAccumulator<QuerySnapshot>();
+          const cacheUnlisten = onSnapshot(
+            query(coll, orderBy('sort', 'asc')),
+            { source: ListenSource.Cache },
+            storeCacheEvent.storeEvent
+          );
+          snapshot = await storeCacheEvent.awaitEvent();
+          expect(toDataArray(snapshot)).to.deep.equal([{ k: 'a', sort: 0 }]);
+          // The metadata is sync with server due to the default listener
+          expect(snapshot.metadata.fromCache).to.equal(false);
```

Comment:
Also trigger an update locally to make sure both get the result.

Reply:
it is tested in "will have synced metadata updates when listening to both cache and default source", where we listen to both source and do local mutation.

Reply:
this and the test below focus on the initialization of multiple listeners with different source, and local mutations are different aspect.